clients/common/norm.cpp

```diff
@@ -506,6 +506,92 @@ double norm_check_general<double>(char        norm_type,
+                                                 rocblas_float_complex* hCPU,
+                                                 rocblas_float_complex* hGPU)
+{
+    // norm type can be O', 'I', 'F', 'o', 'i', 'f' for one, infinity or Frobenius norm
+    // one norm is max column sum
+    // infinity norm is max row sum
+    // Frobenius is l2 norm of matrix entries
+
+    float       work[1];
+    rocblas_int incx  = 1;
+    float       alpha = -1.0f;
+    rocblas_int size  = lda * N;
+
+    double cumulative_error = 0.0;
+
+    for(int i = 0; i < batch_count; i++)
+    {
+        float cpu_norm = clange_(&norm_type, &M, &N, hCPU + i * stride_a, &lda, work);
+
+        caxpy_(&size, &alpha, hCPU + i * stride_a, &incx, hGPU + i * stride_a, &incx);
+
+        float error = clange_(&norm_type, &M, &N, hGPU + i * stride_a, &lda, work) / cpu_norm;
+
+        if(norm_type == 'F' || norm_type == 'f')
+        {
+            cumulative_error += error;
+        }
+        else if(norm_type == 'O' || norm_type == 'o' || norm_type == 'I' || norm_type == 'i')
+        {
+            cumulative_error = cumulative_error > error ? cumulative_error : error;
+        }
+    }
+
+    return cumulative_error;
+}
+
+template <>
+double norm_check_general<rocblas_double_complex>(char                    norm_type,
+                                                  rocblas_int             M,
+                                                  rocblas_int             N,
+                                                  rocblas_int             lda,
+                                                  rocblas_int             stride_a,
+                                                  rocblas_int             batch_count,
+                                                  rocblas_double_complex* hCPU,
+                                                  rocblas_double_complex* hGPU)
+{
+    // norm type can be O', 'I', 'F', 'o', 'i', 'f' for one, infinity or Frobenius norm
+    // one norm is max column sum
+    // infinity norm is max row sum
+    // Frobenius is l2 norm of matrix entries
+
+    double      work[1];
+    rocblas_int incx  = 1;
+    double      alpha = -1.0f;
+    rocblas_int size  = lda * N;
+
+    double cumulative_error = 0.0;
+
+    for(int i = 0; i < batch_count; i++)
+    {
+        double cpu_norm = zlange_(&norm_type, &M, &N, hCPU + i * stride_a, &lda, work);
+
+        zaxpy_(&size, &alpha, hCPU + i * stride_a, &incx, hGPU + i * stride_a, &incx);
+
+        double error = zlange_(&norm_type, &M, &N, hGPU + i * stride_a, &lda, work) / cpu_norm;
+
+        if(norm_type == 'F' || norm_type == 'f')
+        {
+            cumulative_error += error;
+        }
+        else if(norm_type == 'O' || norm_type == 'o' || norm_type == 'I' || norm_type == 'i')
+        {
+            cumulative_error = cumulative_error > error ? cumulative_error : error;
+        }
+    }
+
+    return cumulative_error;
+}
+
```

Comment:
It would be better if most of these norm functions could be consolidated and differentiated where necessary, instead of having the same code multiplied several times.

Reply:
In other words, use templates and make all of the types use the same `norm` code. Then where you need to differentiate between `general`, `rocblas_half`, `rocblas_bf16`, or `rocblas_float_complex`, or `rocblas_double_complex`, you can insert `if` statements (`if constexpr` in C++17) to make code conditional.

Try not to copy-and-paste code. Instead try to refactor existing code to become more general. I can help you with C++ template metaprogramming.